python/grass/pygrass/modules/interface/parameter.py

```diff
@@ -176,7 +176,7 @@ def __init__(self, xparameter=None, diz=None):
             try:
                 # Check for integer ranges: "3-30" or float ranges: "0.0-1.0"
                 isrange = re.match(
-                    "(?P<min>-*\d+.*\d*)*-(?P<max>\d+.*\d*)*", diz["values"][0]
+                    r"(?P<min>-*\d+.*\d*)*-(?P<max>\d+.*\d*)*", diz["values"][0]
```

Comment:
This is probably the first time this line was touched since the introduction of CodeQL check. Follow the "Show more details" link for more information.

Possible solution (not sure it addresses the raised CodeQL issue, but the regex seems to me to work fine):


```suggestion
                    r"(?P<min>-*\d+.*\d*)-(?P<max>\d+.*\d*)", diz["values"][0]
```

Reply:
I want to be sure I understand the wanted regex. Is it only `# Check for integer ranges: "3-30" or float ranges: "0.0-1.0"` that needs to be checked? I'm also looking to understand the `<min>` and `<max>`, and `?P`

Reply:
Original: 
![image](https://github.com/OSGeo/grass/assets/27212526/d889070b-80f5-49fb-964c-8f4539a7c6a4)
Yours:
![image](https://github.com/OSGeo/grass/assets/27212526/1cd5a348-1534-430c-bf9f-b076f8ed87ac)

I'm looking for something better

Reply:
> I want to be sure I understand the wanted regex. Is it only `# Check for integer ranges: "3-30" or float ranges: "0.0-1.0"` that needs to be checked?

Yes, as far as I understand it. I'm not sure what range (pun not intended) of different ranges are expected.

> I'm also looking to understand the `<min>` and `<max>`,

They are *named* groups

>  and `?P`

regex syntax for named groups (https://docs.python.org/3/library/re.html#regular-expression-syntax)

Reply:
Does it need to support scientific notation? bad inputs (multiple dots are rejected)? Min and max reversed? negative values before or after?

Reply:
I think the dot should only appear 0 or once, not 0 or more times. And as written, the dot means any character

Reply:
> Does it need to support scientific notation? bad inputs (multiple dots are rejected)? Min and max reversed? negative values before or after?

Don't think so, haven't so far.

> I think the dot should only appear 0 or once, not 0 or more times. And as written, the dot means any character

Yes, one dot or none at all.

Reply:
And should `.123-76.2`  (no leading 0) work?

What about `0.123-76.` (No trailing 0)

I just found this https://stackoverflow.com/a/12643073, that explains that even an optional dot is not enough, since if its optional, then we end up with a regex like `\d+` and `\d*` back to back, which is another ReDos.

Reply:
```suggestion
                    r"(?P<min>-?(?:\d*\.)?\d+)-(?P<max>-?(?:\d*\.)?\d+)", diz["values"][0]
```
Gives this:
![image](https://github.com/OSGeo/grass/assets/27212526/b5a31cb2-28ec-4e46-a64f-e5fff66e89ca)


But if its okay in this case to say that it **must** match the beginning and end of the string, then we could end up with 

```suggestion
                    r"^(?P<min>-?(?:\d*\.)?\d+)-(?P<max>-?(?:\d*\.)?\d+)$", diz["values"][0]
```
![image](https://github.com/OSGeo/grass/assets/27212526/ec85a0d8-072b-4c5c-950a-89f69d059196)

Reply:
Note "Check for integer ranges" indicates integer may be represented by floats too. This is for parsing input parameters from running modules, I believe, so no need to complicate more than necessary.

Reply:
So this means ? What best fits the flow of process (I don't have on top of my head where this fits on the full sequence that GRASS does from A-Z, I only looked at the local picture here)

Reply:
> Note "Check for integer ranges" indicates integer may be represented by floats too. 

Scratch that.

What I mean is, the regex served well (?!) so far, suppose it expect trivial ranges (eg. not floats like `.33`), but we should focus on the CodeQL issue.

Reply:
So lets try with your first suggestion, and if another CodeQL issue pops up, we'll address it.

Reply:
I hate to unresolve this conversation, but `(?P<min>-*\d+.*\d*)-(?P<max>\d+.*\d*)` does not match `1-` which is what the code seems to account for on line 184. `1-` would be valid input for command line parser definition which is what this is trying to parse. If grass.pygrass should go to to such lengths in parsing and checking the values, that's another discussion (which I just had to revisit recently).

Reply:
What does 1- mean in this context? Is it a valid range for usages in grass, or a test design error? And yes, no problem about unresolving, the conversation resolved itself when committing one of the possibilities. There's also a black fix to apply for the next round of CI.

Reply:
Added with https://github.com/OSGeo/grass/commit/499b5e6a800fdb34fc262622033ac8997b2d36cb “Fix undefined limits in the Parameter class”

Reply:
`(?P<min>[-]{0,1}\d+.*\d*){0,1}-(?P<max>\d+.*\d*){0,1}` ?

Reply:
I think this one, the first of of my comment with two options, is closer, but with a little changes from your suggestion it gives:

```suggestion
                    r"(?P<min>-?(?:\d*\.)?\d+)?-(?P<max>-?(?:\d*\.)?\d+)?", diz["values"][0]
```

I think that there's a logic flaw if this regex I suggest here is what it expected for the tests. Isn't the place where to split between an empty min value ambiguous?

Reply:
I think the requiremet is only "at least one of min, max". See the relevant parser code for ranges of ints and doubles:

https://github.com/OSGeo/grass/blob/04763fd008ece763fe42df6075c4ae0a1fcb98db/lib/gis/parser.c#L1471-L1485

https://github.com/OSGeo/grass/blob/04763fd008ece763fe42df6075c4ae0a1fcb98db/lib/gis/parser.c#L1420-L1434

The above is checking the values provided in command line against what is stored in the definition.

This is used in couple tools:

```bash
$ grep -IrnE 'options.*= "[01]-";' grass grass-addons
grass/display/d.barscale/main.c:122:    length_opt->options = "0-";
grass/raster/r.random.cells/main.c:70:    MaxCells->options = "1-";
grass-addons/src/raster/r.pops.spread/main.cpp:685:    opt.threads->options = "1-";
grass-addons/src/imagery/i.gcp/main.c:91:    opt.point->options = "1-";
$ grep -IrnE "options: [01]-$" grass grass-addons
grass-addons/src/raster/r.sun.daily/r.sun.daily.py:255:# % options: 1-
grass-addons/src/raster/r.sun.hourly/r.sun.hourly.py:233:# % options: 1-
grass-addons/src/raster/r.mapcalc.tiled/r.mapcalc.tiled.py:64:# % options: 1-
grass-addons/src/raster/r.mapcalc.tiled/r.mapcalc.tiled.py:73:# % options: 1-
grass-addons/src/raster/r.shaded.pca/r.shaded.pca.py:87:# % options: 1-
grass-addons/src/raster/r.local.relief/r.local.relief.py:45:# % options: 0-
grass-addons/src/raster/r.lake.series/r.lake.series.py:76:# % options: 0-
grass-addons/src/raster/r.lake.series/r.lake.series.py:95:# % options: 1-
grass-addons/src/raster/r.colors.cubehelix/r.colors.cubehelix.py:58:# % options: 0-
grass-addons/src/raster/r.viewshed.exposure/r.viewshed.exposure.py:153:# % options: 0-
grass-addons/src/raster/r.maxent.lambdas/r.maxent.lambdas.py:123:# % options: 1-
grass-addons/src/vector/v.in.csv/v.in.csv.py:54:# % options: 1-
```

I authored or co-authored eight of those, so clearly there is some documentation problem, what is possible is not what people know about (there is couple of `0-999999` in the code and I remember catching this during a review recently). For the record, I did not author the options range feature. I'm just its happy user.

I'm pretty sure `values` in grass.pygrass is `options` in C and interface definition (_allowed values_ == _available options_).

Reply:
Ok, reading this I think I know what would be equivalent